lib/client-retry.js

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+"use strict";
+const { AssetComputeClient } = require("./client");
+const colors = require("colors/safe");
+const sleep = require("util").promisify(setTimeout);
+
+const MAX_RETRY_429 = 4;
+const MAX_RETRY_WAIT_TIME = 60000;
+const MIN_RETRY_WAIT_TIME = 30000;
+
+/**
+ * Small wrapper around AssetComputeClient adding retry logic on 429s based on `retry-after` header
+ */
+class AssetComputeClientRetry extends AssetComputeClient {
+    constructor(integration, options) {
+        super(integration, options);
+        this.max429RetryCount = (options && options.max429RetryCount) || MAX_RETRY_429;
+    }
+
+    retryWaitTime(e) {
+        // randomly choose between 30-60 seconds waittime
+        let waitTime = Math.floor(Math.random() * (MAX_RETRY_WAIT_TIME - MIN_RETRY_WAIT_TIME + 1)) + MIN_RETRY_WAIT_TIME;
+
+        // in seconds amount of time to wait before retrying the request
+        if (e.retryAfter && typeof(e.retryAfter) === 'number' && !isNaN(e.retryAfter)) {
+            waitTime = e.retryAfter * 1000 + (Math.floor(Math.random() * 1000));
+        }
+        return waitTime;
+    }
+
+    async register() {
+        let response, error;
+        let non429Response = false;
+        let retryCount = 0;
+        while (!non429Response && retryCount <= this.max429RetryCount) {
+            retryCount++;
+            try {
+                response = await super.register();
+                non429Response = true;
+            } catch (e) {
+                if (e.message.includes("429")) {
+                    console.log(colors.red(`encountered a 429: ${e.message}.`));
+                    const waitTime = this.retryWaitTime(e);
+                    console.log(colors.red(`Waiting for ${waitTime}ms and trying again`));
+                    await sleep(waitTime);
+                } else {
+                    non429Response = true;
+                    error = e;
+                }
+            }
+        }
+        if (!non429Response) {
+            throw new Error("Running into 429s. Will not continue testing. Try re-running in a few minutes.");
```

Comment:
In case the library is not used in tests:
```suggestion
            throw new Error("Running into 429s. Will not continue retrying. Try again in a few minutes.");
```